contributors/design-proposals/controller-ref.md

```diff
@@ -1,100 +1,304 @@
-* Extend `OwnerReference` adoption procedure to set a `Controller` field in one of the owners,
-* Update all affected controllers to respect `ControllerRef`.
-
-Necessary related work:
-* `OwnerReferences` are correctly added/deleted,
-* GarbageCollector removes dangling references,
-* Controllers don't take any meaningful actions when `DeletionTimestamps` is set.
-
-# Considered alternatives
+A ControllerRef is thus defined as an OwnerReference with `Controller=true`.
+Each object may have at most one ControllerRef in its list of OwnerReferences.
+The validator for OwnerReferences lists will fail any update that would violate
+this invariant.
+
+# Behavior
+
+This section summarizes the intended behavior for existing controllers.
+It can also serve as a guide for respecting ControllerRef when writing new
+controllers.
+
+## The Three Laws of Controllers
+
+All controllers that manage collections of objects should obey the following
+rules.
+
+1. **Take ownership**
+
+   A controller must claim *ownership* of any objects it creates by adding a
+   ControllerRef, and may also claim ownership of an object it didn't create,
+   as long as the object has no existing ControllerRef (i.e. it is an *orphan*).
+
+1. **Don't interfere**
+
+   A controller may not take any action (e.g. edit/scale/delete) on an object it
+   does not own, except to [*adopt*](#adoption) the object if allowed by the
+   First Law.
+
+1. **Don't share**
+
+   A controller should not count an object it does not own toward satisfying its
+   desired state (e.g. a certain number of replicas), although it may include
+   the object in plans to achieve its desired state (e.g. through adoption)
+   as long as such plans do not conflict with the First or Second Laws.
+
+## Adoption
+
+If a controller finds an orphaned object that matches its selector, it should
+try to adopt the object by adding a ControllerRef.
+
+Multiple controllers may race to adopt a given object, but only one can win
+by being the first to add a ControllerRef to the object's OwnerReferences list.
+The losers will see their adoptions fail due to a validation error as explained
+[above](#api).
+
+If a controller has a non-nil `DeletionTimestamp`, it must not attempt adoption
+or take any other actions except updating its `Status`.
+This prevents readoption of objects orphaned by the [orphan finalizer](garbage-collection.md#part-ii-the-orphan-finalizer)
+during deletion of the controller.
+
+## Orphaning
+
+When a controller is deleted, the objects it owns will either be orphaned or
+deleted according to the normal [Garbage Collection](garbage-collection.md)
+behavior, based on OwnerReferences.
+
+In addition, if a controller finds that it owns an object that no longer matches
+its selector, it should orphan the object by removing itself from the object's
+OwnerReferences list. Since ControllerRef is just a special type of
+OwnerReference, this also means the ControllerRef is removed.
```

Comment:
Typically removal of a label from an object.

Or change of selector (though usually changing selectors is a Bad Idea).